agimus_controller/agimus_controller/mpc.py

```diff
@@ -0,0 +1,63 @@
+import time
+from agimus_controller.ocp_base import OCPBase
+from agimus_controller.warm_start_base import WarmStartBase
+from agimus_controller.trajectory import (
+    TrajectoryPoint,
+    WeightedTrajectoryPoint,
+    TrajectoryBuffer,
+)
+from agimus_controller.mpc_data import MPCResults, MPCDebugData
+
+
+class MPC(object):
+    def __init__(self) -> None:
+        self._ocp = None
+        self._warm_start = None
+        self._mpc_results: MPCResults = None
+        self._mpc_debug_data: MPCDebugData = None
+        self._trajectory_buffer = TrajectoryBuffer()
+
+    def setup(self, ocp: OCPBase, warm_start: WarmStartBase) -> None:
+        self._ocp = ocp
+        self._warm_start = warm_start
+
+    def run(self, initial_state: TrajectoryPoint, current_time_ns: int) -> MPCResults:
+        assert self._ocp is not None
+        assert self._warm_start is not None
+        timer1 = time.perf_counter_ns()
+        self._clear_buffer_past(current_time_ns)
+        reference_trajectory = self._extract_horizon_from_buffer()
+        self._ocp.set_reference_horizon(reference_trajectory)
+        timer2 = time.perf_counter_ns()
+        x0, x_init, u_init = self._warm_start.generate(
+            initial_state, reference_trajectory
+        )
+        timer3 = time.perf_counter_ns()
+        self._ocp.solve(x0, x_init, u_init)
+        timer4 = time.perf_counter_ns()
+
+        # Extract the solution.
+        self._mpc_results = self._ocp.mpc_results
+        self._mpc_debug_data = self._ocp.debug_data
+        self._mpc_debug_data.duration_iteration = timer4 - timer1
+        self._mpc_debug_data.duration_horizon_update = timer2 - timer1
+        self._mpc_debug_data.duration_generate_warm_start = timer3 - timer2
+        self._mpc_debug_data.duration_ocp_solve = timer4 - timer3
+
+    def add_trajectory_point(self, trajectory_point: WeightedTrajectoryPoint):
+        self._trajectory_buffer.append(trajectory_point)
+
+    def add_trajectory_points(self, trajectory_points: list[WeightedTrajectoryPoint]):
+        for trajectory_point in trajectory_points:
+            self.add_trajectory_point(trajectory_point)
+
+    def _clear_buffer_past(self, current_time_ns: int):
```

Comment:
I'm not sure we need something this complicated, because mpc frequency and trajectory discretization frequency are the same, so I think we always have only one point to delete

Reply:
Can you guarantee that MPC will always finish in one loop with a frequency that we set? Computing the very first solution where the warm start is not good might take longer, as far as I know, and that's why we need this logic.

Reply:
It is not true that MPC period and OCP discretization step are the same! They **happen** to have the same value in our demo but that should not be a strong assumption.
For an extreme case, see for instance: https://hal.science/hal-02993058v1/file/ICRA21_skleff.pdf 
MPC period = 1 ms (1kHz)
OCP discretization step = 30 ms (see IV.B.)

Reply:
> Can you guarantee that MPC will always finish in one loop with a frequency that we set? Computing the very first solution where the warm start is not good might take longer, as far as I know, and that's why we need this logic.

Indeed for the first step it might takes longer but nothing starts till it's done and we send the first commands so it's not really a problem (at least it was the case in ros1 and I guess it still is). If the others mpc steps takes more than the time we have for the loop, it means that the mpc is not working as expected so I guess we should assume it's not happening ?



> It is not true that MPC period and OCP discretization step are the same! They **happen** to have the same value in your demo but that should not be a strong assumption. For an extreme case, see for instance: https://hal.science/hal-02993058v1/file/ICRA21_skleff.pdf MPC period = 1 ms (1kHz) OCP discretization step = 30 ms (see IV.B.)

Sure there are some mpc doing this stuff, but since we are doing an mpc to do trajectory tracking, why would we want different frequency for discretization and mpc steps ? Maybe it's something we should talk orally.

Reply:
We could talk about it orally sure, I was probably not very clear in my last message my bad :) Let me rephrase.

There are basically 3 time deltas:
- dt_mpc: between two calls to ocp.solve() basically (decided by the spin frequency of the ros node)
- dt_ocp: discretization step between two nodes of the horizon (fixed at the creation of the OCP)
- dt_plan: discretization step of the plan (coming from HPP, diffusion, oracle...)

I think it is reasonable to assume that dt_ocp = dt_plan. 
However dt_mpc != dt_ocp in the vast majority of cases and I'm affraid assuming this would limit us quite a bit. Not exactly sure if this is what you are suggesting, just want to be clear.

@MaximilienNaveau I'd be interested to have your take on this.

Reply:
Because we're doing trajectory tracking I was thinking to have dt_mpc = dt_ocp as well. 
I was thinking that because if we have for example dt_ocp = 10 ms and dt_mpc = 5 ms, then we will have an mpc iteration between each node, but what will be the references of the nodes in between since we discretized at dt_plan=10ms ? we wouldn't have the right ones, unless we discretize at the same frequency of the mpc and we change all the references, now I'm thinking that it could be useful to give an horizon of references at each mpc step

Reply:
Generally, assuming that we use next point in trajectory and shifting the whole trajectory based on indices will be equivalent to moving them based on timestamps if dt_ocp = dt_traj = dt_mpc. At leat in the case of the case of following the trajectory from HPP.

This stays true, while this implementation lets us extend the buffer to handle cases where dt_ocp != dt_traj.

Reply:
if dt_mpc != dt_ocp why not simply changing the discretization step to have the exact references instead of interpolations ? I think it would make sense to have dt_mpc = dt_plan